src/ORM/EagerLoadedList.php

```diff
@@ -0,0 +1,1003 @@
+                }
+                if (in_array($singleton->getRelationType($relation), ['has_many', 'many_many', 'belongs_many_many'])) {
+                    throw new InvalidArgumentException("$relation is not a linear relation on model $prevModelClass");
+                }
+            }
+
+            if (strpos($column, '.') === false) {
+                if (!singleton($relationModelClass)->hasDatabaseField($column)) {
+                    throw new DatabaseException("Unknown column \"$column\"");
+                }
+                $columnName = '"' . $column . '"';
+            } else {
+                // Find the db field the relation belongs to - It will be returned in quoted SQL "TableName"."ColumnName" notation
+                // Note that sqlColumnForField() throws an expected exception if the field doesn't exist on the relation
+                $relationPrefix = DataQuery::applyRelationPrefix($relations);
+                $columnName = DataObject::getSchema()->sqlColumnForField($relationModelClass, $fieldName, $relationPrefix);
+            }
+
+            // All of the above is necessary to ensure the expected exceptions are thrown for invalid relations
+            // But we still need to ultimately throw an exception here, because sorting by relations isn't
+            // currently supported at all for this class.
+            if (!empty($relations)) {
+                throw new InvalidArgumentException('Cannot sort by relations on EagerLoadedList');
+            }
+        }
+
+        // If $columnName is equal to $col it means that it was orginally raw sql or otherwise invalid.
+        if ($columnName === $column) {
+            throw new InvalidArgumentException("Invalid sort column $column");
+        }
+    }
+
+    public function reverse(): static
+    {
+        // No-op if we're gonna shuffle the list anyway
+        if ($this->sort === 'shuffle') {
+            return $this;
+        }
+        // Set the sort order for each clause to be reversed
+        // This is how DataList reverses its list order as well
+        $list = clone $this;
+        foreach ($list->sort as $clause => &$dir) {
+            $dir = (strtoupper($dir) == 'DESC') ? 'ASC' : 'DESC';
+        }
+        return $list;
+    }
+
+    public function limit(?int $length, int $offset = 0): static
+    {
+        if ($length !== null && $length < 0) {
+            throw new InvalidArgumentException("\$length can not be negative. $length was provided.");
+        }
+
+        if ($offset < 0) {
+            throw new InvalidArgumentException("\$offset can not be negative. $offset was provided.");
+        }
+
+        // We don't actually apply the limit immediately, for compatability with the way it works in DataList
+        $list = clone $this;
+        $list->limitOffset = [$length, $offset];
+        return $list;
+    }
+
+    /**
+     * Check if this list has an item with the given ID
+     */
+    public function has(int $id): bool
```

Comment:
This is a bit of a random method, I'd say probably just delete it since it's not on DataList.  You'd can just just use `byID()` and it for most normal use it'll take the same number of lines?

```php
if ($list->has(99)) {
    $obj = $list->byID(99);
    // do things
}

$obj = $list->byID(99);
if ($obj) {
  // do things
}
```

If you really want to keep it, `has()` isn't very obvious what it does, it should probably be `hasID()`

Reply:
`has()` was added in an earlier refactor of `DataList::fetchEagerLoadRelations()` where we needed to query the list to see if it had the parent record (when the data was still stored in a flat structure rather than a tree).

It looks like somewhere along the way I swapped it out for looping through `getRows()`... which was silly as it is less efficient _and_ required a `break 2` which is a little gross. I've now swapped that back to simply checking `has()` - and renamed that to `hasID()` per your request.

Using `byID()` is a lot less efficient, as it requires instantiating a dataobject, adding its nested eagerloading data, etc - when all we really need to do is check array_key_exists on the rows. Even if we change how we store the rows in the future, having `hasID()` loop over the rows and break when it finds the right one will _still_ be more efficient than instantiating the dataobject.